### DISCUSSION TITLE
`vmap` over a `callable[[float], Array]`

### DISCUSSION THREAD

**Discussion by gautierronan** [jax-ml/jax#20631] [Q&A]
Let's assume I have a function `f` given by the user (i.e. I am allowed to wrap it, but not allowed to modify it or to know its internals). This function has signature `f(t: float) -> Array`. I would like to `jax.vmap` over the returned array of this function.

In other words, I would like to achieve this kind of behavior.
```python
import jax
import jax.numpy as jnp
from jax import Array

def f(t: float) -> Array:
    return jnp.array([t, t**2])

def g(t: float) -> Array:
    return jnp.array([-t**2, -t])

def foo(f: callable[[float], Array], g: callable[[float], Array]) -> Array:
    return f(1.0) + g(2.0)

jax.vmap(foo, in_axes=(0, 0))(f, g)
jax.vmap(foo, in_axes=(0, None))(f, g)
# ValueError: vmap was requested to map its argument along axis 0,
# which implies that its rank should be at least 1, but is only 0 (its shape is ())
```

For standard arrays, it would work fine.
```python
x = jnp.array([1.0, 1.0])
y = jnp.array([-4.0, 2.0])

def bar(x: Array, y: Array) -> Array:
    return x + y

jax.vmap(bar, in_axes=(0, 0))(x, y)
jax.vmap(bar, in_axes=(0, None))(x, y)
# Array([-3.,  3.], dtype=float32)
# Array([[-3.,  3.],
#        [-3.,  3.]], dtype=float32)
```

I'm guessing I should be able to wrap `f` and `g` in a certain way to allow this kind of `vmap`? How could I achieve something like that? I know it should involve `jax.eval_shape` in some way, but would love any kind of pointers to get me started.

Thanks for the help.

**Comment by jakevdp** [comment]
The arguments to a vmapped function must be arrays, not functions. `jax.vmap(foo, in_axes=(0, 0))(f, g)` will not work, because `f` and `g` are functions, not arrays.

The fix would be to do something like your second solution, where you are passing arrays rather than functions as arguments to the vmapped function.

I'm not entirely sure what else to suggest, because it's not clear to me what your desired outcome is when passing a function in place of an array.

**Reply by gautierronan** [reply]
Assuming I could define `f` and `g` inside `foo` (which I don't want to do because it's passed by users in this form for convenience), I could use the following syntax:

```python
def foo(x: Array, y: Array) -> Array:
    def f(t: float) -> Array:
        return t**x
    def g(t: float) -> Array:
        return (-t)**y
    return f(1.0) + g(2.0)

x = jnp.array([1, 2])
y = jnp.array([2, 1])

jax.vmap(foo, in_axes=(0, 0))(x, y)
jax.vmap(foo, in_axes=(0, None))(x, y)
```

However, if I want `f` and `g` to be defined outside of `foo`, then I have to pass `x` and `y` as arguments. e.g. this would work:

```python
def f(t: float, x: Array) -> Array:
    return t**x
def g(t: float, y: Array) -> Array:
    return (-t)**y

def foo(x: Array, y: Array) -> Array:
    return f(1.0, x) + g(2.0, y)

x = jnp.array([1, 2])
y = jnp.array([2, 1])

jax.vmap(foo, in_axes=(0, 0))(x, y)
jax.vmap(foo, in_axes=(0, None))(x, y)
```

Is it possible to push this logic even further, and to avoid `x` and `y` as extra args inside `f` and `g`? And thus to "vmap" over `f` and `g`? I was thinking of doing a bit of magic, for instance with custom PyTrees. Something like:

```python
@jax.register_pytree_node_class
class PyTreeCallable:
    def __init__(self, f: callable[[float], Array]):
        self.f = f
        self.shape = jax.eval_shape(f, 0.0)

    def tree_flatten(self):
        children = [LeafCallable(self.f, i) for i in range(self.shape[0])]
        return children, None

    @classmethod
    def tree_unflatten(cls, aux_data, children):
        return cls(lambda t: jnp.stack([c.f(t) for c in children]))

class LeafCallable:
    def __init__(self, f: callable[[float], Array], i: int):
        self.f = f
        self.i = i

    def __call__(self, t: float) -> Array:
        return self.f(t)[self.i]

vmappable_f = PyTreeCallable(f)
vmappable_g = PyTreeCallable(g)
```

But I'm looking for suggestions in this regard, because I'm quite sure this is not the proper way to do it. Thanks a lot for your help.

**Reply by jakevdp** [reply]
If I'm understanding correctly, you want to define `f` and `g` without reference to `x` and `y`, but somehow have them operate on `x` and `y`? Leaving `vmap` aside, can you show me a Python program with the kind of semantics you want? It sounds like you're saying you want something like this:
```
def f(t):
  return x + t

def foo(f, x):
  return f(2)
```
Where the `x` within `foo` somehow is referenced within `f`, without explicitly passing the argument. Am I understanding correctly?

**Reply by gautierronan** [reply]
Yes exactly, I'm thinking of something with the following form:
```python
x = 3.0
def f(t):
    return x + t

def foo(f):
    return f(2)
```

In the case I'm looking at, the function `foo` has `f` as input and `f` has a fixed output shape (which can be evaluted using `jax.eval_shape`). So I'm guessing that in principle, `vmap` has all the knowledge required to vectorize over the output dimension of `f`. It doesn't actually need to know about `x`, if that makes sense.

**Reply by jakevdp** [reply]
There's no way to do that kind of action-at-a-distance in Python, so there's no way to do that kind of action-at-a-distance in JAX. You could maybe hack things by temporarily setting global variables within the outer function, but that seems brittle and confusing at the very best.

I'd suggest defining your functions to take explicit arguments rather than relying on global state.

**Reply by gautierronan** [reply]
I don't quite agree. What I'm looking for is not action-at-a-distance. The following jax syntax is very much valid (i.e. fully explicit) and cannot be parametrized by an additional explicit argument in any simple way. Yet, it should be vmappable in the sense that it's just the lowered code executing things at once instead of sequentially. 

```python
def f(t: float) -> Array:
    return jnp.array([t, 3 * t, t**2])

def foo(f: callable[[float], Array]) -> Array:
    return f(1.0) + f(2.0)
```

This example in particular will run without any need for a `vmap`. But my very first example wouldn't.

**Reply by jakevdp** [reply]
OK, I guess I didn't understand what you wanted to do after all.

If what you want is for `1.0` and `2.0` to be arrays, you'll have to create those arrays and pass them as arguments to `f` and `g`. There's no way to implicitly convert constant scalars to arrays using `vmap`.

**Reply by jakevdp** [reply]
For example, this will work:
```python
def f(t: float) -> Array:
    return jnp.array([t, 3 * t, t**2])

def foo(x, y, f: Callable[[float], Array]) -> Array:
    return f(x) + f(y)

jax.vmap(foo, in_axes=(0, 0, None))(jnp.ones(3), jnp.zeros(3), f)
```
Or, if you want to map the inner function rather than the outer, you can do this:
```python
foo(jnp.ones(3), jnp.zeros(3), jax.vmap(f))
```
